src/VisualStudio/LiveShare/Impl/DiagnosticHandler.cs

```diff
@@ -0,0 +1,90 @@
+
+using System;
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.LanguageServer;
+using Microsoft.VisualStudio.LanguageServices.LiveShare.CustomProtocol;
+using Microsoft.VisualStudio.LiveShare.LanguageServices;
+using Microsoft.VisualStudio.Threading;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.VisualStudio.LanguageServices.LiveShare
+{
+    /// <summary>
+    /// Handles a request to get diagnostics. This is a custom request that is issued on document
+    /// open. This type also provides notifications when diagnostics change.
+    /// TODO - Move implementation to lower LSP layer once we figure out how to deal with notify / client interactions.
+    /// </summary>
+    internal class DiagnosticsHandler : ILspRequestHandler<TextDocumentParams, LSP.Diagnostic[], Solution>, ILspNotificationProvider
+    {
+        private readonly IDiagnosticService _diagnosticService;
+
+        public event AsyncEventHandler<LanguageServiceNotifyEventArgs> NotifyAsync;
+
+        [ImportingConstructor]
+        public DiagnosticsHandler(IDiagnosticService diagnosticService)
+        {
+            _diagnosticService = diagnosticService ?? throw new ArgumentNullException(nameof(diagnosticService));
+            _diagnosticService.DiagnosticsUpdated += DiagnosticService_DiagnosticsUpdated;
+        }
+
+        private async void DiagnosticService_DiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs e)
+        {
+            // Since this is an async void method, exceptions here will crash the host VS. We catch exceptions here to make sure that we don't crash the host since
+            // the worst outcome here is that guests may not see all diagnostics.
+            try
+            {
+                // LSP doesnt support diagnostics without a document. So if we get project level diagnostics without a document, ignore them.
+                if (e.DocumentId != null && e.Solution != null)
+                {
+                    var document = e.Solution.GetDocument(e.DocumentId);
+                    if (document == null || document.FilePath == null)
+                    {
+                        return;
+                    }
+
+                    var lspDiagnostics = await GetDiagnosticsAsync(e.Solution, document, CancellationToken.None).ConfigureAwait(false);
+                    var publishDiagnosticsParams = new LSP.PublishDiagnosticParams { Diagnostics = lspDiagnostics, Uri = document.GetURI() };
+                    var eventArgs = new LanguageServiceNotifyEventArgs(LSP.Methods.TextDocumentPublishDiagnosticsName, publishDiagnosticsParams);
+                    await (NotifyAsync?.InvokeAsync(this, eventArgs)).ConfigureAwait(false);
+                }
+            }
+            catch (Exception)
```

Comment:
let's do our default pattern here where we check cancellation and then do NFW
catch(Exception ex) when (FatalError.XXX)

Reply:
I meant ReprotWithoutCrashAndNotPropagate